trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracestate
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+const (
+	keyMaxSize       = 256
+	valueMaxSize     = 256
+	maxKeyValuePairs = 32
+)
+
+const (
+	keyWithoutVendorFormat = `[a-z][_0-9a-z\-\*\/]{0,255}`
+	keyWithVendorFormat    = `[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}`
+	keyFormat              = `(` + keyWithoutVendorFormat + `)|(` + keyWithVendorFormat + `)`
+	valueFormat            = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var keyValidationRegExp = regexp.MustCompile(`^(` + keyFormat + `)$`)
+var valueValidationRegExp = regexp.MustCompile(`^(` + valueFormat + `)$`)
+
+// Tracestate represent tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	tracestateEntries []*TracestateEntry
+}
+
+// TracestateEntry represent one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type TracestateEntry struct {
+	key   string
+	value string
+}
+
+// Key returns the key of TracestateEntry
+func (te *TracestateEntry) Key() string {
+	return te.key
+}
+
+// Value returns the value of TracestateEntry
+func (te *TracestateEntry) Value() string {
+	return te.value
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then false is returned with the value "" is returned.
+// If the key is found then true is returned with its value.
+func (ts *Tracestate) Get(key string) (string, bool) {
+	if len(ts.tracestateEntries) == 0 {
+		return "", false
+	}
+	for _, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			return entry.value, true
+		}
+	}
+	return "", false
+}
+
+// TraceEntries returns a slice of TracestateEntry.
+func (ts *Tracestate) TraceEntries() []*TracestateEntry {
+	return ts.tracestateEntries
+}
+
+func (ts *Tracestate) remove(key string) {
+	for index, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			ts.tracestateEntries = append(ts.tracestateEntries[:index], ts.tracestateEntries[index+1:]...)
+			break
+		}
+	}
+}
+
+func (ts *Tracestate) add(entry *TracestateEntry) error {
+	ts.remove(entry.Key())
+	if len(ts.tracestateEntries) >= maxKeyValuePairs {
+		return fmt.Errorf("Set failed: reached maximum key/value pairs limit of %d", maxKeyValuePairs)
```

Comment:
We might need an API for truncating the tracestate (removing key-value-pairs) from the tail.
Consider the following scenario:
1. I got a tracestate with 31 key-value-pairs (which is maxKeyValuePairs  - 1).
2. The trace library I'm using would want to add 2 additional key-value-pairs.
3. What the library would want to do, is to **remove the last key-value-pair from the existing tracestate**, and prepend the 2 additional key-value-pairs.

There might be other scenarios, for example, remove the key-value-pairs from the tail until the total size (depending on the serialization detail) is under a certain limit (e.g. "The length of a combined header MUST be less than or equal to 512 bytes. If the length of a combined header is more than 512 bytes it SHOULD be ignored.").

Reply:
Maybe add 'truncate' as a bool argument to NewXXX. If set then remove pairs from the tail.

#2 - would require that NewXXX takes more than one key-value pair. Alternatively. you can call NewXXX repeatedly but that will not be efficient.

Regarding 512 bytes - should that be handled in propagation or should it be part of truncation?

Do you want me to add the above to this PR or open up a new one?

Reply:
We can leave it for the next PR since it's not a blocking issue.

The 512 bytes limit is specific to the HTTP header format, which should not be part of the tracestate itself.

I mention it here to discuss if we need a "truncate" function. (do we want to propagator to truncate the key-value-pairs by calling a tracestate method, or directly access the key-value-pairs and discard from the tail?)

Reply:
@reyang can you explain the use case of needing to add 2 additional key-value pairs? 

Typically a tracing system should only encode a single key-value pair where key is their vendor label and value typically holds (embeds) multiple values important to that vendor's propagation format. 

Example: Zipkin old style B3 has multiple key-value pairs as separate headers but in W3C's tracestate it will use single value (thus multi embedded values in header value): `b3={x-b3-traceid}-{x-b3-spanid}-{if x-b3-flags 'd' else x-b3-sampled}-{x-b3-parentspanid}`

Also see spec example of 2 different tracing systems as it highlights single k/v per tracing system: https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md#examples-of-http-headers-1

Reply:
@basvanbeek I was thinking about the migration scenario:
1. You have service A, B and C. A and B are using a legacy protocol X for tracing, while B and C use another one Y.
2. B and C are owned by you (which means you can change the code and do deployments), while A is not (e.g. external service).
3. By upgrading the protocol between B and C to W3C distributed tracing, you want to carry the existing information in Tracestate during the transition period.
4. This would end up with X and Y in the Tracestate header for requests issued from service B to C.